### PR TITLE
Limit OOMScoreAdj when running in UserNS ("Rootful-in-Rootless")

### LINKED ISSUE #46563: Rootful-in-Rootless dind doesn't work since systemd v250 (due to oom score adj): `runc create failed: unable to start container process: can't get final child's PID from pipe: EOF: unknown.`
### Description

Rootful-in-Rootless dind doesn't work since systemd v250 (Dec 2021): 

```console
$ docker run -d --name dind --privileged docker:24.0.6-dind

$ docker exec dind docker run --rm hello-world
docker: Error response from daemon: failed to create task for container: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: can't get final child's PID from pipe: EOF: unknown.
```

This is because systemd v250 bumped up the default oom scores for user sessions:
- https://github.com/systemd/systemd/pull/20893

### Reproduce

See above

### Expected behavior

It should work

### docker version (parent)

```bash
Client: Docker Engine - Community
 Version:           24.0.6
 API version:       1.43
 Go version:        go1.20.7
 Git commit:        ed223bc
 Built:             Mon Sep  4 12:31:40 2023
 OS/Arch:           linux/amd64
 Context:           rootless

Server: Docker Engine - Community
 Engine:
  Version:          24.0.6
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.7
  Git commit:       1a79695
  Built:            Mon Sep  4 12:31:40 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.24
  GitCommit:        61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc:
  Version:          1.1.9
  GitCommit:        v1.1.9-0-gccaecfc
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
 rootlesskit:
  Version:          1.1.1
  ApiVersion:       1.1.1
  NetworkDriver:    slirp4netns
  PortDriver:       builtin
  StateDir:         /tmp/rootlesskit396465062
 slirp4netns:
  Version:          1.2.0
  GitCommit:        656041d45cfca7a4176f6b7eed9e4fe6c11e8383
```

### docker info (parent)

```bash
Client: Docker Engine - Community
 Version:    24.0.6
 Context:    rootless
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.11.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.21.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 1
  Running: 1
  Paused: 0
  Stopped: 0
 Images: 2
 Server Version: 24.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: false
  userxattr: true
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 61f9fd88f79f081d64d6fa3bb1a0dc71ec870523
 runc version: v1.1.9-0-gccaecfc
 init version: de40ad0
 Security Options:
  seccomp
   Profile: builtin
  rootless
  cgroupns
 Kernel Version: 6.2.0-33-generic
 Operating System: Ubuntu 23.04
 OSType: linux
 Architecture: x86_64
 CPUs: 4
 Total Memory: 3.809GiB
 Name: lima-docker
 ID: 58287749-998e-420b-8a8f-93b483ed5102
 Docker Root Dir: /home/suda.linux/.local/share/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

## Workaround

Set `--oom-score-adj=100` in the child `docker run` (when the container's `oom_score_adj` is set to 200)

```console
$ docker exec dind cat /proc/1/oom_score_adj
200

$ docker exec dind docker run --rm --oom-score-adj=100 hello-world

Hello from Docker!
...
```

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**- What I did**

Fix #46563  `Rootful-in-Rootless dind doesn't work since systemd v250 (due to oom score adj)`

**- How I did it**

By limiting OOMScoreAdj

**- How to verify it**

Run `BINDDIR=. make shell` with Rootless Docker and systemd >= 250, and run `docker run --rm hello-world` inside there

**- Description for the changelog**

Fix "Rootful-in-Rootless" dind on systemd >= 250

**- A picture of a cute animal (not mandatory but encouraged)**
🐧

### CODE REVIEW

**Comment by thaJeztah** [comment]
Per the discussion on https://github.com/moby/moby/issues/46563#issuecomment-1740030857

> > I'm slightly confused; do we set an oom-score-adjust by default? (And if not, does that mean that it now must be set always)?
>
> Set to `0` by default, and there is no way to unset it currently, as `HostConfig` can't distinguish zero value from "unset".
>
> https://github.com/moby/moby/blob/e2c23ed57de3bc768b928d0350b9c65d03d45de2/api/types/container/hostconfig.go#L442

Would there be other possible approaches to this problem? I _THINK_ the _intent_ has always been for `OOMScoreAdj: 0` to be "don't change the score, and use default". Before systemd's change that default would be (IIUC) `0`, so "don't change" effectively worked "by accident" (as we were setting it to the value that also was the default), but that part broke because of the systemd change?

So if we want to adjust the behavior to the original _intent_, what options would we have for that? Should we change it to a pointer (if we consider `0` to be a value we want to allow being explicitly set: although I'm not sure if the API and CLI would currently do this correctly), or just consider `0` to be "not set" (and don't adjust the score)?

**Comment by thaJeztah** [comment]
FWIW, I recall looking at moving the WithOOMScoreAdj options to containerd, as it didn't provide these, but it turned out that containerd has a (somewhat artificial) split between options that are implemented in CRI, and options in the "general" options. OOMScore options _are_ available in containerd, but they're defined in the CRI package. Perhaps that's something we should look at as well, and move more of those to be "generic" options (which both CRI and non-CRI uses could use); https://github.com/containerd/containerd/blob/c5ce2a5f750ad89ab02974d047744517346578f3/pkg/cri/opts/spec_linux_opts.go#L385C1-L449

**Comment by AkihiroSuda** [comment]
> So if we want to adjust the behavior to the original intent, what options would we have for that? Should we change it to a pointer (if we consider 0 to be a value we want to allow being explicitly set: although I'm not sure if the API and CLI would currently do this correctly), or just consider 0 to be "not set" (and don't adjust the score)?

This is worth considering for vNext, but doesn't seem backportable.

**Comment by AkihiroSuda** [comment]
Context: I need this PR to support running cri-dockerd in minikube with Rootless Docker driver (`minikube start --rootless --driver=docker --container-runtime=docker`)

https://github.com/AkihiroSuda/minikube/commit/ec95bcfcbe9602006ff81f93a5600c15a509fcbc

File: pkg/rootless/specconv/specconv_linux.go [hunk 1]
```diff
@@ -13,6 +13,18 @@ import (
...
+// the daemon is running in UserNS but has no access to RootlessKit API socket, host filesystem, etc.
+//
+// This fuction does:
+// * Fix up OOMScoreAdj
+func ToRootfulInRootless(spec *specs.Spec) {
+	currentOOMScoreAdj := getCurrentOOMScoreAdj()
+	if spec.Process != nil && spec.Process.OOMScoreAdj != nil && *spec.Process.OOMScoreAdj < currentOOMScoreAdj {
+		*spec.Process.OOMScoreAdj = currentOOMScoreAdj
+	}
+}
```
Comment (thaJeztah) [inline pkg/rootless/specconv/specconv_linux.go:26]: Micro-nit; I guess we can skip `getCurrentOOMScoreAdj()` if we don't need it;

```go
func ToRootfulInRootless(spec *specs.Spec) {
	if spec.Process == nil || spec.Process.OOMScoreAdj == nil {
		return
	}
	if currentOOMScoreAdj := getCurrentOOMScoreAdj() && *spec.Process.OOMScoreAdj < currentOOMScoreAdj {
		*spec.Process.OOMScoreAdj = currentOOMScoreAdj
	}
}
```

File: pkg/rootless/specconv/specconv_linux.go [hunk 2]
```diff
@@ -13,6 +13,18 @@ import (
 	specs "github.com/opencontainers/runtime-spec/specs-go"
 )
 
+// ToRootfulInRootless is used for "rootful-in-rootless" dind;
+// the daemon is running in UserNS but has no access to RootlessKit API socket, host filesystem, etc.
+//
+// This fuction does:
+// * Fix up OOMScoreAdj
```
Comment (thaJeztah) [inline pkg/rootless/specconv/specconv_linux.go:20]: Is it worth leaving a reference here why we're doing this? I'm sure that information gets easily lost.

File: daemon/oci_linux.go [hunk 3]
```diff
@@ -1126,6 +1135,8 @@ func (daemon *Daemon) createSpec(ctx context.Context, daemonCfg *configStore, c
 	}
 	if daemonCfg.Rootless {
 		opts = append(opts, withRootless(daemon, &daemonCfg.Config))
+	} else if userns.RunningInUserNS() {
+		opts = append(opts, withRootfulInRootless(daemon, &daemonCfg.Config))
```
Comment (thaJeztah) [inline daemon/oci_linux.go:1139]: quite orthogonal to this PR (sorry for continue asking things - I should do some further digging myself); so the situation with "rootful" in "rootless" is that it doesn't _allow_ setting a score lower than the environment it's running in; but I'm wondering now if it's _desirable_ to allow this in the "non-rootless" situation, i.e. currently this is allowed;

Run a docker-in-docker container with a `500` oom-score adjust;

```bash
docker run -it --rm --privileged -v /var/lib/docker --oom-score-adj 500 docker:dind sh
/# dockerd --debug
```

Within the container, run a container with a oom-score adjust of `100` 

```bash
cat /proc/self/oom_score_adj
500

docker run --rm --oom-score-adj 100 busybox cat /proc/self/oom_score_adj
100
```

In this case (technically) the outer container would be killable before the busybox container (of course, the busybox container would be gone together with it, so it's more of a "theoretical" example), just wondering if we should ultimately go the length, and produce a sensible error for this situation.

Comment (AkihiroSuda) [reply daemon/oci_linux.go:1139]: > but I'm wondering now if it's desirable to allow this in the "non-rootless" situation, i.e. currently this is allowed;

I guess a warning should be printed, but producing an error may break somebody's environment

File: pkg/rootless/specconv/specconv_linux.go (hunk 1 above)
Comment (AkihiroSuda) [reply pkg/rootless/specconv/specconv_linux.go]: Fixed

File: pkg/rootless/specconv/specconv_linux.go (hunk 2 above)
Comment (AkihiroSuda) [reply pkg/rootless/specconv/specconv_linux.go]: Fixed

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
@neersighted ptal

**Review by neersighted — APPROVED** [review]

**Pull request merged**